Zend/zend_ast.c

```diff
@@ -702,6 +702,41 @@ ZEND_API zend_result ZEND_FASTCALL zend_ast_evaluate_inner(
 			}
 			zval_ptr_dtor_nogc(&op1);
 			break;
+		case ZEND_AST_CAST:
+			if (UNEXPECTED(zend_ast_evaluate_ex(&op1, ast->child[0], scope, &short_circuited, ctx) != SUCCESS)) {
+				ret = FAILURE;
+				break;
+			}
+			if (ast->attr == Z_TYPE(op1)) {
+				ZVAL_COPY_VALUE(result, &op1);
+			} else {
+				switch (ast->attr) {
+					case _IS_BOOL:
+						ZVAL_BOOL(result, zend_is_true(&op1));
+						break;
+					case IS_LONG:
+						ZVAL_LONG(result, zval_get_long_func(&op1, false));
+						break;
+					case IS_DOUBLE:
+						ZVAL_DOUBLE(result, zval_get_double_func(&op1));
+						break;
+					case IS_STRING:
+						ZVAL_STR(result, zval_get_string_func(&op1));
+						break;
+					case IS_ARRAY:
+						zend_cast_zval_to_array(result, &op1, IS_VAR);
+						break;
+					case IS_OBJECT:
+						zend_cast_zval_to_object(result, &op1, IS_VAR);
+						break;
+					EMPTY_SWITCH_DEFAULT_CASE();
```

Comment:
what about the new `(void)` cast?

Reply:
Void casts are not allowed in expression context, only when their return value is not used. This never applies to constant expressions. This is also a different ast type (`ZEND_AST_CAST_VOID`), so it doesn't need to be handled here.